Adding CCT bypass docs (Use locale-specific domains without displaying them) (#1605)
MicrosoftDocs/edge-developer: microsoft-edge/progressive-web-apps-chromium/how-to/microsoft-store.md

description: Make your PWA more discoverable by publishing in the Microsoft Store.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.prod: microsoft-edge
ms.technology: pwa
ms.date: 12/01/2021
keywords: progressive web apps, PWA, Edge, Windows, Microsoft Store
---

> # Publish your Progressive Web App to the Microsoft Store

Reviewer: I might resize/shrink the screenshot in section "Create an app reservation"

### Submit your app package to the Store

To submit your app to the Microsoft Store:

1.  Go to [Windows Partner Center](https://partner.microsoft.com/dashboard/windows/overview)
1.  Select your app.
1.  Select **Start your Submission**.

    :::image type="content" source="../media/windows-partner-center-start-submission.msft.png" alt-text="Start a new app submission on Windows Partner Center.":::

1.  When you're prompted, provide information about your app, such as pricing and age rating.

1.  On the **Packages** prompt, select the `.msixbundle` and the `.classic.appxbundle` files you generated in the [Package your PWA](#package-your-pwa-for-the-store) section.

After you complete your submission, your app is reviewed, typically within 24 to 48 hours.  After you receive approval, your PWA is available in the Microsoft Store.

<!-- ====================================================================== -->
## Measure usage of your PWA installed from the Microsoft Store

> When your PWA is initially launched, if the PWA was installed from the Microsoft Store, Microsoft Edge includes the following `Referer` header with the request for the first navigation of your web app:

Reviewer 2: ```suggestion
When your PWA is initially launched, if the PWA was installed from the Microsoft Store, Microsoft Edge includes the following `Referer` header with the request of the first navigation of your web app.
```

```
Referer: app-info://platform/microsoft-store
```

Use this feature to measure distinct traffic from your PWA that was installed from the Microsoft Store.  Based on the traffic, you can adjust your app's content to improve the user experience.  This feature is accessible to both client and server code. To access this information on the client side, you can query `document.referrer` in your JavaScript.

This feature was first introduced in Microsoft Edge version 91, and the DOM API was introduced in Microsoft Edge version 93.

<!-- ====================================================================== -->
## Redirect to locale-specific domains without displaying additional UI

By default, a PWA that's installed from the Microsoft Store displays an additional UI when the app is redirected to a locale-specific domain.  The added UI shows the URL and page title.  This UI is added because navigation to the locale-specific domain is considered "out-of-scope".  However, you can prevent this UI from being displayed, by specifying locale-specific origins that are associated with the PWA.

> The following figure shows the UI that is introduced when a user moves outside the scope of a PWA: 

Reviewer 2: ```suggestion
The following figure shows the UI that is introduced when a user moves outside the scope of a PWA. 
```

Displaying the URL and page title is a security feature to ensure that users know they have left the context of the PWA.  This added UI makes sense when a user loads a page from another website in the context of the PWA.  However, that added UI may be inappropriate when a user moves among domains that are all part of the same website.

### Prevent the locale-specific URL and title from being displayed

To prevent the additional UI from being shown in a PWA that's installed from the Microsoft Store, you can use [URL Handlers](https://github.com/WICG/pwa-url-handler/blob/main/explainer.md) to enable the PWA to span multiple locale-specific domains.

To prevent displaying the URL and page title:

1. Within the PWA's Web App Manifest, use [the `url_handlers` member](https://github.com/WICG/pwa-url-handler/blob/main/explainer.md#manifest-changes) to specify an array of origins that are associated with that app.

> 1. On each of the referenced origins, include a [`web-app-origin-association` file](https://github.com/WICG/pwa-url-handler/blob/main/explainer.md#web-app-origin-association-file) that associates the PWA with that domain.

Reviewer 2: @aarongustafson - in the linked example, is Example 1's intro sentence explaining the JSON that follows it? It's unclear to me in Example 1 what "web-app-origin-association file" refers to.

Author: Yes. Example 1 shows what the web-app-origin-association file (which is JSON) should look like in each domain.

This feature was first introduced in Microsoft Edge version 91, and the DOM API was introduced in Microsoft Edge version 93.

<!-- ====================================================================== -->

> ## Created a unified experience across multiple domains

Reviewer 2: ```suggestion
## Use multiple locale-specific domains in PWAs
```

> A Web App Manifest is tied to a single domain, but some PWAs use locale-specific domains for their customers in specific regions of the globe. When visiting the PWA in a web browser, customers are seamlessly transitioned from the principal domain (e.g., contoso.com) to a locale-specific one (e.g., contoso.co.ke) because the redirect happens during initial load of that website. Customers who install the PWA from Edge would therefor install it from the locale-specific domain and all future launches of that PWA will go directly to that domain instead of the principal one.
> 

Reviewer 2: ```suggestion
A Web App Manifest is tied to a single domain, but some PWAs use locale-specific domains for their customers in specific regions of the globe. When visiting the PWA in a web browser, customers are seamlessly transitioned from the principal domain (for example, contoso.com) to a locale-specific one (for example, contoso.co.ke) because the redirect happens during initial load of that website. Customers who install the PWA from Edge would therefore install it from the locale-specific domain and all future launches of that PWA will go directly to that domain instead of the principal one.

```

> Store-installed PWAs, however, have a hard-coded start URL that is pointed at the principal domain. When the PWA is launched, it navigates to the principal domain and then a customer may (as necessary) be redirected to their locale-specific domain. If that redireciton occurs, the navgation is considered "out of scope" and the app UI will display the URL and title at the top of the web page. This is a security feature to ensure users are aware that they have left the context of the app. That makes sense when users are loading a page from another website in the context of the PWA, but can be a little jarring when users are moving between domains that are all part of the same app.

Reviewer: ```suggestion
Store-installed PWAs, however, have a hard-coded start URL that is pointed at the principal domain. When the PWA is launched, it navigates to the principal domain and then a customer may (as necessary) be redirected to their locale-specific domain. If that redirection occurs, the navigation is considered "out of scope" and the app UI will display the URL and title at the top of the web page. This is a security feature to ensure users are aware that they have left the context of the app. That makes sense when users are loading a page from another website in the context of the PWA, but can be a little jarring when users are moving between domains that are all part of the same app.
```
I plan to make such fixes in a Writer/Editor pass.  quick note for now

Reviewer 2: Changed app to website in the last sentence.

Reviewer 2: @aarongustafson - can we add a screenshot of the security UI that displays when the user is redirected. That will help land the before and after picture to highlight the change that we're documenting in this PR.

Developers can use [URL Handlers](https://github.com/WICG/pwa-url-handler/blob/main/explainer.md) to enable their Store PWAs to span multiple locale-specific domains without the additional UI being shown. There are two steps to enabling a PWA to handle navigations for multiple domains. First, within the PWA’s Web App Manifest, [the `url_handlers` member](https://github.com/WICG/pwa-url-handler/blob/main/explainer.md#manifest-changes) is used to indicate an array of origins that should be associated with that app. Then, on each of the referenced origins, developers must include [a `web-app-origin-association` file](https://github.com/WICG/pwa-url-handler/blob/main/explainer.md#web-app-origin-association-file) that attests to the PWA’s association with that domain. With these in place, Edge will no longer show the additional UI when the principal domain is redirected to the locale-specific domains it is set up to handle.

> Eventually, the `url_handlers` feature will be replaced by [`scope_extensions`](https://github.com/WICG/manifest-incubations/blob/gh-pages/scope_extensions-explainer.md), but that spec is still in deveopment. Once finalized, this same behavior will apply for domains included within an app’s extended scope.

Reviewer 2: ```suggestion
Eventually, the `url_handlers` feature will be replaced by [`scope_extensions`](https://github.com/WICG/manifest-incubations/blob/gh-pages/scope_extensions-explainer.md), but that spec is still in development. Once finalized, this same behavior will apply for domains included within an app’s extended scope.
```

## Use locale-specific domains without displaying them

By default, store-installed PWAs display an unexpected additional UI showing the URL and page title when the app is redirected to a locale-specific domain.  This happens because the navigation to the locale-specific domain is considered "out-of-scope".  However, you can prevent this UI from being displayed, by specifying locale-specific origins that are associated with the PWA.

### Domain redirection with browser-installed PWAs

A Web App Manifest is tied to a single domain.  However, some PWAs use locale-specific domains for their customers in specific regions of the globe.  When visiting the PWA in a web browser, customers are seamlessly transitioned from the principal domain (for example, contoso.com) to a locale-specific domain (for example, contoso.co.ke), because the redirect happens during initial load of that website.

Customers who install the PWA from Microsoft Edge would therefore install the PWA from the locale-specific domain.  Subsequent launches of the PWA go directly to that locale-specific domain, instead of first going to the principal domain.

### Domain redirection with store-installed PWAs

> Store-installed PWAs, however, have a hard-coded start URL that is pointed at the principal domain.  When the PWA is launched, the PWA initially navigates to the principal domain, and then a customer may (as necessary) be redirected to their locale-specific domain. If that redirection occurs, the navigation is considered "out of scope" and the app UI will display the URL and title at the top of the web page.

Reviewer: "at the top of the web page" - in the titlebar?  can we clarify the description or add a screenshot?

This is a security feature to ensure that users are aware that they have left the context of the app.  That makes sense when a user is loading a page from another website in the context of the PWA, but may be inappropriate when a user moves among domains that are all part of the same website.

### Prevent the locale-specific URL and title from being displayed

> To prevent the additional UI from being shown, you can use [URL Handlers](https://github.com/WICG/pwa-url-handler/blob/main/explainer.md) to enable the Store PWA to span multiple locale-specific domains.  There are two steps to enabling a PWA to handle navigations for multiple domains:

Reviewer: is the phrase "the Store PWA" correct?

<!-- ====================================================================== -->
## Redirect to locale-specific domains without displaying additional UI

By default, a PWA that's installed from the Microsoft Store displays an additional UI when the app is redirected to a locale-specific domain.  The added UI shows the URL and page title.  This UI is added because navigation to the locale-specific domain is considered "out-of-scope".  However, you can prevent this UI from being displayed, by specifying locale-specific origins that are associated with the PWA.

The following figure shows the UI that is introduced when a user moves outside the scope of a PWA:

> :::image type="content" source="../media/locale-redirection-additional-ui.png" alt-text="The additional UI with URL and page title when the app is redirected to a locale-specific domain.":::

Author: a locale-specific domain -> another domain.
